src/ResultWriter/DefaultResultWriter.php

```diff
@@ -67,41 +70,34 @@ public function writeToCsv(QueryResult $result, ExportConfig $exportConfig, stri
 
         // Write rows
         try {
-            $this->writeRows($exportConfig, $iterator, $csvWriter);
-        } catch (NoRowsException $e) {
-            @unlink($csvFilePath); // no rows, no file
-            return new ExportResult($csvFilePath, 0, $this->getMaxValueFromState($exportConfig));
+            $this->writeRows($iterator, $result->getMetadata(), $exportConfig, $csvWriter);
         } finally {
             $result->closeCursor();
         }
 
-        $incFetchingColMaxValue = $this->getIncrementalFetchingMaxValue($exportConfig, $this->lastRow);
+        $incFetchingColMaxValue = $this->lastRow ?
+            $this->getIncrementalFetchingMaxValue($exportConfig, $this->lastRow) :
+            $this->getMaxValueFromState($exportConfig);
         return new ExportResult($csvFilePath, $this->rowsCount, $incFetchingColMaxValue);
     }
 
-    protected function writeRows(ExportConfig $exportConfig, Iterator $iterator, CsvWriter $csvWriter): void
-    {
-        // No rows found ?
-        if (!$iterator->valid()) {
-            throw new NoRowsException();
-        }
-
+    protected function writeRows(
+        Iterator $iterator,
+        QueryMetadata $queryMetadata,
+        ExportConfig $exportConfig,
+        CsvWriter $csvWriter
+    ): void {
         // With custom query are no metadata in manifest, so header must be present
         $includeHeader = $exportConfig->hasQuery();
 
-        // Rows found, iterate!
-        $resultRow = $iterator->current();
-        $iterator->next();
-
-        // Write header and first line
+        // Write header
         if ($includeHeader) {
-            $this->writeHeader(array_keys($resultRow), $csvWriter);
+            $this->writeHeader($queryMetadata->getColumns()->getNames(), $csvWriter);
         }
-        $this->writeRow($resultRow, $csvWriter);
 
         // Write the rest
-        $this->rowsCount = 1;
-        $this->lastRow = $resultRow;
+        $this->rowsCount = 0;
```

Comment:
Prvy riadok sa uz nespracovava oddelene.

Reply:
super, tím se to tu dost zjednodušilo 👍